File: lib/rate-limit.ts
type SlidingWindowRule = {
  windowMs: number;
  maxRequests: number;
};

type RateLimitResult =
  | { allowed: true }
  | { allowed: false; retryAfterSeconds: number };

const globalForRateLimit = globalThis as typeof globalThis & {
  __appointmentRateLimitStore?: Map<string, number[]>;
};

const rateLimitStore =
  globalForRateLimit.__appointmentRateLimitStore ?? new Map<string, number[]>();

if (!globalForRateLimit.__appointmentRateLimitStore) {
  globalForRateLimit.__appointmentRateLimitStore = rateLimitStore;
}

const DEFAULT_RULES: SlidingWindowRule[] = [
  { windowMs: 60_000, maxRequests: 6 },
  { windowMs: 3_600_000, maxRequests: 30 },
];

export function checkRateLimit(
  key: string,
  rules: SlidingWindowRule[] = DEFAULT_RULES
): RateLimitResult {
  const now = Date.now();
  const maxWindowMs = Math.max(...rules.map((rule) => rule.windowMs));
  const existingHits = rateLimitStore.get(key) ?? [];
  const recentHits = existingHits.filter((ts) => now - ts < maxWindowMs);

  for (const rule of rules) {
    const hitsInWindow = recentHits.filter((ts) => now - ts < rule.windowMs);
    if (hitsInWindow.length >= rule.maxRequests) {
      const oldestHitInWindow = hitsInWindow[0];
      const retryAfterSeconds = Math.max(
        1,
        Math.ceil((rule.windowMs - (now - oldestHitInWindow)) / 1000)
      );
      rateLimitStore.set(key, recentHits);
      return { allowed: false, retryAfterSeconds };
    }
  }

  recentHits.push(now);
  rateLimitStore.set(key, recentHits);
  return { allowed: true };
}

File: backend/core/rate_limit.py
from __future__ import annotations

import time
from dataclasses import dataclass


@dataclass
class SlidingWindowRule:
    window_ms: int
    max_requests: int


DEFAULT_RULES = [
    SlidingWindowRule(window_ms=60_000, max_requests=6),
    SlidingWindowRule(window_ms=3_600_000, max_requests=30),
]

_store: dict[str, list[float]] = {}


def check_rate_limit(key: str, rules: list[SlidingWindowRule] | None = None) -> tuple[bool, int | None]:
    now = time.time() * 1000
    active_rules = rules or DEFAULT_RULES
    max_window = max(rule.window_ms for rule in active_rules)
    recent_hits = [ts for ts in _store.get(key, []) if now - ts < max_window]

    for rule in active_rules:
        in_window = [ts for ts in recent_hits if now - ts < rule.window_ms]
        if len(in_window) >= rule.max_requests:
            oldest = in_window[0]
            retry_after = max(1, int((rule.window_ms - (now - oldest)) / 1000))
            _store[key] = recent_hits
            return False, retry_after

    recent_hits.append(now)
    _store[key] = recent_hits
    return True, None
